packages/flutter/lib/src/material/color_scheme.dart

```diff
@@ -209,39 +418,113 @@ class ColorScheme with Diagnosticable {
   /// The color displayed most frequently across your app’s screens and components.
   final Color primary;
 
-  /// A darker version of the primary color.
-  final Color primaryVariant;
+  /// A color that's clearly legible when drawn on [primary].
+  ///
+  /// To ensure that an app is accessible, a contrast ratio of 4.5:1 for [primary]
+  /// and [onPrimary] is recommended. See
+  /// <https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html>.
+  final Color onPrimary;
+
+  final Color? _primaryContainer;
+  /// A color used for elements needing less emphasis than [primary].
+  Color get primaryContainer => _primaryContainer ?? primary;
```

Comment:
AFAIK we should not fallback for any of these. There's nothing in spec about it, and it adds confusion to the color system

Reply:
I think we do want to fallback for folks who might try to use the new colors without providing them, since the properties are non-null. Otherwise, we'd have to make them nullable and then half the colors would non-null and half nullable.

Reply:
As a developer, I would find it really confusing if I did `colorScheme.primaryContainer` and it silently returned `colorScheme.primary`. It gets worse at the component level, where understanding why component X uses color Y now requires understanding the fallbacks for certain color roles. I think this sort of optimisation is ultimately more confusing than helpful. 

> Otherwise, we'd have to make them nullable and then half the colors would non-null and half nullable.

What's wrong with that?

Reply:
Yeah, this is a compromise for making this change non-breaking.  The problem here is that we can't make them non-null parameters like the rest of the colors because we would break existing apps that didn't provide them. If we make them fully nullable then every usage of it will need to handle a null color which is something you don't have to deal with in the rest of the API. It would be confusing to have to remember which colors need to be handled differently. And every usage might handle the null color in a different way. This way it is at least consistent when it is wrong and is hopefully only slightly wrong (i.e. primary is at least close to primaryContainer).

If the developer is confused by `colorScheme.primaryContainer` returning `primary`, they can provide a value for `primaryContainer` when they create the color scheme. Or they can use the new generator which will provide one.

Reply:
Thanks for explaining. It's a tricky situation indeed.. Just throwing this idea, what if the default was `Color.red` in the same way that text defaults to red without a `Material` ancestor? That would give a clear signal to developers that they should specify the color, i.e. in cases where they don't use the generator. Moreover, it would help us identify any issues with component theming (much easier to red than a similar color).

Reply:
Interesting idea. I need to think about it some more. 

My initial reaction is that it would yield bad results for people that are updating their apps to M3 without an updated ColorScheme (i.e. they have a color scheme they are already using that doesn't have all the new colors). They would wind up with red-on-red Floating Action buttons, etc. While this is useful to show that there is a problem, it would not be a great experience out of the box. In addition, it is not obvious what the problem is, or what they need to do to fix it.

But I will spend some time exploring this idea. In the mean time, I updated the `ColorScheme` constructor with more info about the null-able colors. Let me know what you think.

Reply:
Although it would be nice to visually alert developers to the need for updating the ColorScheme, I think we can handle that in the M3 migration guide. I agree that making the "out of the box" experience as smooth as we can is important.